lib/ecto/changeset/relation.ex

```diff
@@ -510,6 +512,17 @@ defmodule Ecto.Changeset.Relation do
     end
   end
 
+  defp change_pk(pks) do
+    fn %Changeset{} = cs ->
+      Enum.map(pks, fn pk ->
+        case cs.changes do
+          %{^pk => pk_value} -> pk_value
+          _ -> Map.get(cs.data, pk)
+        end
+      end)
+    end
+  end
```

Comment:
It looks to me like anything getting to `changeset.changes` has its key normalized to atom and value casted to the right type. But please correct me if I'm wrong.

Reply:
Correct!